apps/nsq_to_file/file_logger.go

```diff
@@ -151,123 +151,138 @@ func (f *FileLogger) router(r *nsq.Consumer) {
+	err = f.out.Close()
+	if err != nil {
+		f.logf(lg.FATAL, "failed to close output file: %s", err)
+		os.Exit(1)
+	}
 
-			// Optimistic rename
-			log.Printf("INFO: moving finished file %s to %s", src, dst)
-			err := atomicRename(src, dst)
-			if err == nil {
-				return
-			} else if !os.IsExist(err) {
-				log.Fatalf("ERROR: unable to move file from %s to %s: %s", src, dst, err)
-				return
-			}
+	// Move file from work dir to output dir if necessary, taking care not
+	// to overwrite existing files
+	if f.opts.WorkDir != f.opts.OutputDir {
+		src := f.out.Name()
+		dst := filepath.Join(f.opts.OutputDir, strings.TrimPrefix(src, f.opts.WorkDir))
+
+		// Optimistic rename
+		f.logf(lg.INFO, "moving finished file %s to %s", src, dst)
+		err := exclusiveRename(src, dst)
+		if err == nil {
+			return
+		} else if !os.IsExist(err) {
+			f.logf(lg.FATAL, "unable to move file from %s to %s: %s", src, dst, err)
+			os.Exit(1)
+		}
 
-			// Optimistic rename failed, so we need to generate a new
-			// destination file name by bumping the revision number.
-			_, filenameTmpl := filepath.Split(f.lastFilename)
-			dstDir, _ := filepath.Split(dst)
-			dstTmpl := filepath.Join(dstDir, filenameTmpl)
+		// Optimistic rename failed, so we need to generate a new
+		// destination file name by bumping the revision number.
+		_, filenameTmpl := filepath.Split(f.filename)
+		dstDir, _ := filepath.Split(dst)
+		dstTmpl := filepath.Join(dstDir, filenameTmpl)
 
-			for i := f.rev + 1; ; i++ {
-				log.Printf("INFO: destination file already exists: %s", dst)
-				dst := strings.Replace(dstTmpl, "<REV>", fmt.Sprintf("-%06d", i), -1)
-				err := atomicRename(src, dst)
-				if err != nil {
-					if os.IsExist(err) {
-						continue // next rev
-					}
-					log.Fatalf("ERROR: unable to rename file from %s to %s: %s", src, dst, err)
-					return
+		for i := f.rev + 1; ; i++ {
+			f.logf(lg.WARN, "destination file already exists: %s", dst)
+			dst := strings.Replace(dstTmpl, "<REV>", fmt.Sprintf("-%06d", i), -1)
+			err := exclusiveRename(src, dst)
+			if err != nil {
+				if os.IsExist(err) {
+					continue // next rev
 				}
-				log.Printf("INFO: renamed finished file %s to %s to avoid overwrite", src, dst)
-				break
+				f.logf(lg.FATAL, "unable to rename file from %s to %s: %s", src, dst, err)
+				os.Exit(1)
 			}
+			f.logf(lg.INFO, "renamed finished file %s to %s to avoid overwrite", src, dst)
+			break
 		}
-
-		f.out = nil
 	}
+
+	f.out = nil
 }
 
-func (f *FileLogger) Write(p []byte) (n int, err error) {
-	f.filesize += int64(len(p))
-	return f.out.Write(p)
+func (f *FileLogger) Write(p []byte) (int, error) {
+	n, err := f.writer.Write(p)
+	f.filesize += int64(n)
+	return n, err
 }
 
 func (f *FileLogger) Sync() error {
-	var err error
 	if f.gzipWriter != nil {
-		f.gzipWriter.Close()
-		err = f.out.Sync()
-		f.gzipWriter, _ = gzip.NewWriterLevel(f, f.compressionLevel)
-		f.writer = f.gzipWriter
-	} else {
-		err = f.out.Sync()
+		err := f.gzipWriter.Flush()
```

Comment:
I think i'm 👎on this switch, but I acknowledge that it is possible to encounter unexpected issues with a stream of indipendent gzip chunks (i.e. equiv to `cat file1.gz file2.gz`).

Without separate gzip streams, recovering from a corrupt file (kill -9 w/ partial data written but not sync'd) becomes problematic. 

I think it comes down to the fact that Flush does not write a gzip footer which has the checksum. Close does.

Reply:
Unless I’m misunderstanding something, the recoverable section would be equivalent in either case (bound by the configured sync interval)?

You’re suggesting that by writing out _closing_ footers somehow more data would be recoverable? Given gzip’s streamable quality, I assume that a complete block (regardless of a closed stream) would be decompressable?

Reply:
> I assume that a complete block (regardless of a closed stream) would be decompressable?

decompressable yes, error checkable no. The footer contains the final crc for a stream which you don't get from a sync()

https://www.forensicswiki.org/wiki/Gzip#File_footer

### recoverability

With separate streams for each logical sync, I've found it easy to step through the streams of a file (using [`compress/gzip.Reader.Multistream`](https://golang.org/pkg/compress/gzip/#Reader.Multistream) and use the valid streams to recover a corrupt file. When you don't have a stream boundary recovery is less precise because you can't disambiguate between written but not sync'd data and written and sync'd data. That increases the risk of including records that were never ack'd back to nsqd which isn't desirable.

Reply:
Nice, that feels like a strong argument to revert, thanks.